<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Изучение-данных" data-toc-modified-id="Изучение-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Изучение данных</a></span></li><li><span><a href="#Решение-заданий" data-toc-modified-id="Решение-заданий-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Решение заданий</a></span><ul class="toc-item"><li><span><a href="#Определить-количество-книг,-вышедших-после-1-января-2000-года" data-toc-modified-id="Определить-количество-книг,-вышедших-после-1-января-2000-года-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Определить количество книг, вышедших после 1 января 2000 года</a></span></li><li><span><a href="#Определить-количество-обзоров-и-среднюю-оценку-по-каждой-книге" data-toc-modified-id="Определить-количество-обзоров-и-среднюю-оценку-по-каждой-книге-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Определить количество обзоров и среднюю оценку по каждой книге</a></span></li><li><span><a href="#Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Определить издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Определить-автора-с-самой-высокой-средней-оценкой-книг-—-учитывать-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определить-автора-с-самой-высокой-средней-оценкой-книг-—-учитывать-только-книги-с-50-и-более-оценками-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками</a></span></li><li><span><a href="#Определить-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="Определить-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Определить среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li></ul></li><li><span><a href="#Выводы-по-каждой-из-решённых-задач" data-toc-modified-id="Выводы-по-каждой-из-решённых-задач-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы по каждой из решённых задач</a></span></li></ul></div>

# Анализ базы данных (SQL)

**Цель проекта:** Провести анализ имеющейся базы данных по заранее заданным параметрам. Результаты анализа помогут сформулировать ценностное предложение для нового продукта, связанного с сервисом для чтения книг по подписке. 
    

**Исходные данные** — база данных с информацией о книгах, издательствах, авторах, а также пользовательские обзоры книг.


**План работы по проекту:**<p>  
1. Исследовать таблицы — вывести первые строки.  
2. Сделать по одному SQL-запросу для решения каждого задания:  
- Посчитать, сколько книг вышло после 1 января 2000 года;  
- Для каждой книги посчитать количество обзоров и среднюю оценку;  
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так будут исключены из анализа брошюры;  
- Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;  
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.  
`Вывести результаты каждого запроса в тетрадке.`<br>      
  
<br>  
3. Привести выводы по каждой из решённых задач.  
    

      


## Изучение данных

In [1]:
#импорт библиотек
import pandas as pd
from sqlalchemy import create_engine
from datetime import datetime

In [2]:
# создаем подключение к базе данных, загружаем таблицу и выгружаем данные в файл:
# Задаём параметры подключения к БД
db_config = {'user': '***************', # имя пользователя
            'pwd': '**************', # пароль
            'host': '********************.mdb.yandexcloud.net',
            'port': ****, # порт подключения
            'db': '*********-final-project-db'} # название базы данных

# Формируем строку соединения с БД
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# Подключаемся к БД
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [3]:
# Формируем sql-запрос вывод первых строк таблицы books.
query = ''' SELECT *
            FROM books
            LIMIT 5
        '''
#посмотрим результат запроса
pd.io.sql.read_sql(query, con = engine) 

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
# Формируем sql-запрос вывод первых строк таблицы authors.
query = ''' SELECT *
            FROM authors
            LIMIT 5
        '''
#посмотрим результат запроса
pd.io.sql.read_sql(query, con = engine) 

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
# Формируем sql-запрос вывод первых строк таблицы ratings.
query = ''' SELECT *
            FROM ratings
            LIMIT 5
        '''
#посмотрим результат запроса
pd.io.sql.read_sql(query, con = engine) 

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [40]:
# Формируем sql-запрос вывод первых строк таблицы reviews.
query = ''' SELECT *
            FROM reviews
            LIMIT 5
        '''
      
#посмотрим результат запроса
pd.io.sql.read_sql(query, con = engine) 

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [7]:
# Формируем sql-запрос вывод первых строк таблицы publishers.
query = ''' SELECT *
            FROM publishers
            LIMIT 5
        '''
#посмотрим результат запроса
pd.io.sql.read_sql(query, con = engine) 

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Данные по всем указанным в задании таблицам получены, выведены для ознакомления. 

## Решение заданий

### Определить количество книг, вышедших после 1 января 2000 года

In [8]:
# Формируем sql-запрос для решения задачи по таблицам ratings, reviews.
query = ''' SELECT COUNT (title)                                           -- названия книг по условию даты публикации 
            FROM books
            WHERE DATE_TRUNC('day', publication_date)::date > '2000-01-01' -- берем дату после 01.01.2000
        '''
#посмотрим результат запроса
pd.io.sql.read_sql(query, con = engine) 

count
0    819

После первого января 2000 года вышло 819 книг.

### Определить количество обзоров и среднюю оценку по каждой книге

In [41]:
#Посчитаем количество ревью в базе данных
query = '''SELECT COUNT (review_id) AS review_CNT      
           FROM reviews
        '''
#посмотрим результат запроса
pd.io.sql.read_sql(query, con = engine)  

review_cnt
0        2793

In [58]:
#Проверка среднего рейтинга книги Twilight (Twilight #1)
query = ''' SELECT AVG (rating) AS AVG_rating      --Twilight (Twilight #1)
            FROM ratings
            WHERE book_id=948
        '''
#посмотрим результат запроса
pd.io.sql.read_sql(query, con = engine) 


avg_rating
0      3.6625

In [60]:
#Формируем sql-запрос для решения задачи по таблицам books, reviews, ratings.
query = '''
    SELECT b.book_id, b.title, 
           COUNT (DISTINCT rv.review_id) AS review_cnt, 
           ROUND(AVG (rt.rating), 2) AS avg_rating              -- в сгруппированном списке выводим данные
    FROM books as b
    LEFT JOIN reviews AS rv ON b.book_id = rv.book_id
    LEFT JOIN ratings AS rt ON b.book_id = rt.book_id            -- обьединили в 1 таблицу все данные
    GROUP BY b.book_id                                          -- группируем по индексу книги  
    ORDER BY review_cnt DESC                -- сортируем группировку по индексу книги по количеству обзоров
    '''
#посмотрим результат запроса
pd.io.sql.read_sql(query, con = engine) 

book_id                                              title  review_cnt  \
0        948                            Twilight (Twilight  #1)           7   
1        963                                Water for Elephants           6   
2        734                                   The Glass Castle           6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...           6   
4        695  The Curious Incident of the Dog in the Night-Time           6   
..       ...                                                ...         ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel           0   
996      808                            The Natural Way to Draw           0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites           0   
998      221                          Essential Tales and Poems           0   
999      191  Disney's Beauty and the Beast (A Little Golden...           0   

     avg_rating  
0          3.66  
1          3.98  
2          4.21  
3          4.41  
4          4.08  
..          ...  
995        3.67  
996        3.00  
997        5.00  
998        4.00  
999        4.00  

[1000 rows x 4 columns]

In [61]:
# Проверка по полю количество review после решения задачи.
query = '''
WITH TOTAL AS (
    SELECT b.book_id, b.title, 
           COUNT (DISTINCT rv.review_id) AS review_cnt, 
           ROUND(AVG (rt.rating), 2) AS avg_rating              -- в сгруппированном списке выводим данные
    FROM books as b
    LEFT JOIN reviews AS rv ON b.book_id = rv.book_id
    LEFT JOIN ratings AS rt ON b.book_id = rt.book_id            -- обьединили в 1 таблицу все данные
    GROUP BY b.book_id                                          -- группируем по индексу книги  
    ORDER BY review_cnt DESC)              -- сортируем группировку по индексу книги по количеству обзоров
    
    SELECT SUM (review_CNT)
    FROM TOTAL
    '''
#посмотрим результат запроса
pd.io.sql.read_sql(query, con = engine) 

sum
0  2793.0

Получили сведения о количестве обзоров каждой книги и ее среднем рейтинге.
Всего 1000 книг с оценками и обзорами. Наибольшее количество обзоров у книги с названием Twilight (Twilight #1)	и средним рейтингом 3,66.

In [ ]:
# Формируем sql-запрос для решения задачи по таблицам books, reviews, ratings.
query = '''
    SELECT b.title, 
           COUNT (rv.review_id) AS review_cnt, 
           ROUND(AVG (rt.rating), 2) AS avg_rating               -- в сгруппированном списке выводим данные
    FROM books as b
    LEFT JOIN reviews AS rv ON b.book_id = rv.book_id
    LEFT JOIN ratings AS rt ON b.book_id = rt.book_id            -- обьединили в 1 таблицу все данные
    GROUP BY b.book_id                                          -- группируем по индексу книги  
    ORDER BY review_cnt DESC                                    -- сортируем группировку по индексу книги по количеству обзоров
    '''
#посмотрим результат запроса
pd.io.sql.read_sql(query, con = engine) 

### Определить издательство, которое выпустило наибольшее число книг толще 50 страниц

In [10]:
# Формируем sql-запрос для решения задачи по таблице books, publishers.
query = '''
    SELECT b.publisher_id, 
           p.publisher, 
           COUNT (b.book_id) AS book_count                          -- в сгруппированном списке выводим данные
    FROM books AS b
    LEFT JOIN publishers AS p ON p.publisher_id = b.publisher_id
    WHERE b.num_pages > 50                                          --сначала выберем книги толще 50 страниц
    GROUP BY b.publisher_id, p.publisher                      -- группируем по издательствам, которые такие книги выпустило
    ORDER BY book_count DESC                              -- сортируем группировку по количеству выбранных книг по убыванию
    LIMIT 1
        '''
#посмотрим результат запроса
pd.io.sql.read_sql(query, con = engine) 

publisher_id      publisher  book_count
0           212  Penguin Books          42

Наибольшее число книг толще 50 страниц выпустило издание Penguin Books - 42 таких книги.

### Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками

In [11]:
# Формируем sql-запрос для решения задачи по таблице authors, ratings.
query = '''
    WITH Books_50 AS (
        SELECT book_id, 
               ROUND (AVG (rating), 2) AS rating, 
               COUNT (rating) AS rating_CNT
        FROM ratings 
        GROUP BY book_id
        HAVING COUNT (rating) > 50
    )                                                       -- записываем список книг с 50 и более оценками в таблицу Books_50
    
    SELECT a.author_id,                                     -- в сгруппированном списке выводим данные
           a.author,
           ROUND (AVG (rt.rating), 2) AS avg_rating, 
           SUM (rt.rating_CNT) AS rating_CNT,
           COUNT (rt.book_id) AS book_CNT
    FROM books AS b
    LEFT JOIN authors AS a ON a.author_id = b.author_id
    JOIN Books_50 AS rt ON rt.book_id = b.book_id         -- обьединили в 1 таблицу все данные для книг с 50 и более оценками
    GROUP BY a.author_id                                  -- группируем по авторам, которые такие книги выпустили
    ORDER BY avg_rating DESC                              -- сортируем группировку по сред.рейтингу авторов по убыванию
    LIMIT 1
    '''
#посмотрим результат запроса
pd.io.sql.read_sql(query, con = engine) 

author_id                      author  avg_rating  rating_cnt  book_cnt
0        236  J.K. Rowling/Mary GrandPré        4.29       310.0         4

Имя автора с самой высокой средней оценкой книг, если учитывать только книги с 50 и более оценками, - это J.K. Rowling/Mary GrandPré, имеющий среднюю оценку по книгам 4.29, и общее количество оценок по 4 книгам - 310.

### Определить среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [12]:
# Формируем sql-запрос для решения задачи по таблицам reviews, ratings.
query = '''
    WITH Users_50 AS (
            SELECT username, COUNT (review_id) AS review_CNT     -- в сгруппированном списке выводим данные о количестве 
            FROM reviews                                         -- обзоров у пользователей 
            WHERE username IN (
                    SELECT username                             -- отбираем пользователейс с более 50 оценок
                    FROM ratings
                    GROUP BY username
                    HAVING COUNT (rating_id) > 50
                   ) 
            GROUP BY username                                   -- группируем по пользователям для вывода списка
     )
     
     SELECT ROUND (AVG (review_CNT), 0) AS AVG_review_CNT    -- среднее количество обзоров среди отобранных пользователей
     FROM Users_50                                           -- с округлением до целого.
        '''

#посмотрим результат запроса
pd.io.sql.read_sql(query, con = engine)              

avg_review_cnt
0            24.0

Среднее количество обзоров среди пользователей, которые поставили более 50 оценок, с округлением до целого, составляет 24 обзора.

## Выводы по каждой из решённых задач

1. Данные по всем указанным в задании таблицам получены, выведены для ознакомления.  <br>
<p>
2. Составлены SQL-запросы по задачам и получены следующие данные:  <br>
<p>
    2.1  Задача: Определить количество книг, вышедших после 1 января 2000 года.<br>   
     Данные: После первого января 2000 года вышло 819 книг.  <p>
    2.2  Задача: Определить количество обзоров и среднюю оценку по каждой книге.  <br>
     Данные: Получили сведения о количестве обзоров каждой книги и ее среднем рейтинге.
Всего 1000 книг с оценками и обзорами. Наибольшее количество обзоров у книги с названием Twilight (Twilight #1)	и средним рейтингом 3,66.  <p>
    2.3  Задача: Определить издательство, которое выпустило наибольшее число книг толще 50 страниц. <br> 
     Данные: Наибольшее число книг толще 50 страниц выпустило издание Penguin Books - 42 таких книги. <p> 
    2.4  Задача: Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками.  <br>
     Данные: Имя автора с самой высокой средней оценкой книг, если учитывать только книги с 50 и более оценками, - это J.K. Rowling/Mary GrandPré, имеющий среднюю оценку по книгам 4.29, и общее количество оценок по 4 книгам - 310.  <p>
    2.5  Задача: Определить среднее количество обзоров от пользователей, которые поставили больше 50 оценок.<br>      Данные: Среднее количество обзоров среди пользователей, которые поставили более 50 оценок, с округлением до целого, составляет 24 обзора.<p>